### ISSUE TITLE
Provider behavior differs from kubectl: unable to specify empty resource attribute

### ISSUE THREAD

**Issue by dak1n1** [hashicorp/terraform-provider-kubernetes#872]
### Terraform Version and Provider Version

Terraform v0.12.26
terraform-provider-kubernetes v1.11.3

### Affected Resource(s)

`kubernetes_persistent_volume_claim` is the most widely reported, but the issue is at the SDK level and therefore affects all resources.

### Terraform Configuration Files

This is one example of trying to specify an empty value for the storageclass.

```hcl
resource "kubernetes_persistent_volume" "example" {
  metadata {
    name = "nfs-volume"
  }
  spec {
    capacity = {
      storage = "1T"
    }
    storage_class_name = ""
    access_modes = ["ReadWriteMany"]
    persistent_volume_source {
      nfs {
        server = google_filestore_instance.test.networks[0].ip_addresses[0]
        path = "/${google_filestore_instance.test.file_shares[0].name}"
      }
    }
  }
}

resource "kubernetes_persistent_volume_claim" "example" {
  metadata {
    name = "mariadb-data"
    namespace = "test"
  }
  spec {
    access_modes = ["ReadWriteMany"]
    storage_class_name = ""
    volume_name = kubernetes_persistent_volume.example.metadata[0].name
    resources {
      requests = {
        storage = "1T"
      }
    }
  }
}
```

### Expected Behavior

Users should be able to specify an empty storageclass (`storage_class = ""`), which is behavior supported by kubectl.

### Actual Behavior

The provider interprets `""` to be the same as omitting the value altogether. [There is no way to differentiate between null and empty values](https://discuss.hashicorp.com/t/differentiate-between-null-and-empty-value/8134) in the current version of the SDK.

### Steps to Reproduce

1. Create a PVC with an empty storage class.
2. `terraform apply`

When `terraform apply` has finished, the storageclass is now set to `default` instead of empty (`""`). This behavior is unexpected because it differs from kubectl, subsequently breaking users' workflows.

### References

The recommended work-around is to create a storageclass and use it in PVs and PVCs, instead of using `""` for the storageclass name. An example of that work-around can be found [here](https://github.com/terraform-providers/terraform-provider-kubernetes/tree/master/_examples/google-gke-nfs-filestore).

Other issues and PRs relating to this will be closed, so we can track them all in this single issue.
* https://github.com/terraform-providers/terraform-provider-kubernetes/pull/590
* https://github.com/terraform-providers/terraform-provider-kubernetes/issues/339
* https://github.com/terraform-providers/terraform-provider-kubernetes/issues/864
* https://github.com/terraform-providers/terraform-provider-kubernetes/issues/102
* https://github.com/terraform-providers/terraform-provider-kubernetes/issues/800
* https://github.com/terraform-providers/terraform-provider-kubernetes/issues/339
* https://github.com/terraform-providers/terraform-provider-kubernetes/issues/567
* https://github.com/terraform-providers/terraform-provider-kubernetes/pull/865

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by josephmilla** [comment]
@dak1n1 I'm also facing this issue. Created this PR for allowing empty strings for the `storage_class_name` attribute - https://github.com/hashicorp/terraform-provider-kubernetes/pull/923. Seems to be straightforward, please take a look.

**Comment by taylorturner** [comment]
Any update on when this might be fixed?

When defining `storage_class_name = ""` OR commenting the line out on a persistent_volume resource then looking at `kubectl get pv` it shows the resource as belonging to "blank" storage class BOTH ways.

When doing the same for a persistent_volume_claim resource, either defining `storage_class_name = ""` or commenting it out, kubectl shows the resource belonging to the "default" storage class BOTH ways. 

This makes it so the PVC will not come up because the storage_class_name doesn't match the PV.

Major bug!!

**Comment by mbarrien** [comment]
@taylorturner See workaround documented in the PR References section. This one won't be fixed quickly.

**Comment by dak1n1** [comment]
Since version 2 of the Terraform SDK has been released, I followed up about potentially being able to support passing an empty value into a provider attribute (like `storage_class_name=""`). Unfortunately, this is not something the SDK can support. There is a lower-level SDK that can support the differentiation between null and empty, which I'll put here for reference:  https://github.com/hashicorp/terraform-plugin-go

But we don't have any plans for the Kubernetes provider to use that SDK, since that would be a re-write of the entire provider.

The easiest work-around is to create a storage class (there is an [example here](https://github.com/hashicorp/terraform-provider-kubernetes/tree/master/_examples/google-gke-nfs-filestore)). An alternative work-around, if you don't want to create a storage class, would be to use the [Kubernetes alpha provider](https://github.com/hashicorp/terraform-provider-kubernetes-alpha) to create the PVC, which would allow you to write out the persistent volume claim similarly to yaml, where you could specify the empty storage class.

**Comment by dak1n1** [comment]
I re-opened this one since we're still getting PRs for the feature. I think the work-arounds are not sufficient. And with some information I recently learned about the SDK, implementing the empty storageclass might be possible. It's worth investigating further.

**Comment by dak1n1** [comment]
The easiest work-around in the mean time is to look at your cluster's default storage class (`kubectl get storageclass`) and then use the name of that storage class in your PVC and PV. Here's an example: https://gist.githubusercontent.com/dak1n1/6a95719c4e68cfc9e736c8a7624f53a1/raw/939f038b810d314a689bb687ddfbe674aacc69e6/gistfile1.txt

**Comment by Madmaxguy** [comment]
> The easiest work-around in the mean time is to look at your cluster's default storage class (`kubectl get storageclass`) and then use the name of that storage class in your PVC and PV. Here's an example: https://gist.githubusercontent.com/dak1n1/6a95719c4e68cfc9e736c8a7624f53a1/raw/939f038b810d314a689bb687ddfbe674aacc69e6/gistfile1.txt

This does not help if you need to mount an NFS volume which requires storageClass to be actually empty so the volume does not get provisioned and instead just mounts it

**Comment by dak1n1** [comment]
Here's the NFS work-around in case anyone missed that:

https://github.com/hashicorp/terraform-provider-kubernetes/blob/main/_examples/google-gke-nfs-filestore/main.tf#L130-L172

There shouldn't be any scenarios in which a blank storageclass is literally required, but it is certainly more convenient to use a blank rather than creating your own storageclass and using that.

But the work to support the blank storageclass is finally on our roadmap, so I'm hoping to have support for that sometime in the next 6 months (assuming it ends up being possible to do with the terraform SDK).

**Comment by amarkevich** [comment]
AKS comes with `storage_class_name  = "azurefile"`. Use this in both `kubernetes_persistent_volume` & `kubernetes_persistent_volume_claim`

**Comment by yaleman** [comment]
Not-stale

**Comment by yaleman** [comment]
Nuh uh

**Comment by iamasmith** [comment]
I'm going to supercede the previous post which was what I believed was going on having re-created the successfully working PV/PVC pair.

In order to arrive back at a 'working' PV and PVC combination I ended up having to remove the PV as well as the PVC.

What was actually going on was that the PVC creation was indeed not honoring the empty value and choosing the default storage class which is standard-rwo in this case... with a VolumeBindingMode: WaitForFirstConsumer.

In this mode the PVC does not try to provision the volume using the storageclass until it is used, avioding allocation until necessary and fits in with the reclaim capabilities associated with the volume that one supposes it eventually provisions.

When the PV is applied _at the same time_ the PVC still uses that default storageclass, however, the PV defintion causes the PV on NFS to bind to it because it has the claimRef. The PVC is now satisfied and will use the PV even though it doesn't match the default storageclass.

Furthermore, now the volume is bound to the claim the volume values are used for storage operations and whilst the PVC actual storageclass says Reclaim: Delete, the PV is set to Retain so is kept if the PVC is removed. This is good.

The fact that the PVC now has a mismatch between the drivers it thinks are in use is now irrelevant as the PV is in place and the only think possible is the PVC calling provisioner code using the storageclass it believes is in general use. Those values will not match the PV so the PV is not at risk of being broken.

Basically one _can_ work with this if they are confident they know all the issues associated with this mismatch in configuration.

In my scenario where I may need to shrink the NFS volume all I need to do is remember to remove BOTH the PV and the PVC from my config an re-add them to allow the reprovisioning of the PVC and the bound state created by the properly configured PV.

**Comment by iamasmith** [comment]
I looked at this further and understand the issue appears to be related in part to the SDK and possibly another discussion that is ongoing as an Issue that the SDK in some, possibly earlier iterations, has persisted what should have been null values as empty strings and providers may be catering for this by string comparison and not setting the associated structure member.

There is a workaround of course that might help here, given this field identifies a resource by a name that must fit within certain constraints it would appear to be trivial to be able to provide a specific text that intentionally breaks the scheme, possibly something like "<EXPLICITLY_EMPTY>" that could, as a result of the string match then set a truly empty string for the field.

It's not ideal but avoids the rather awkward issue ongoing in the SDK discussions around what to do with providers that have been persisting null values as empty strings.
